C error when using two generics

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: macos, macOS, 11.6, 20G165
Processor: 16 cpus, 64bit, little endian, Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
CC version: Apple clang version 13.0.0 (clang-1300.0.29.3)

getwd: /Users/wes/dev/test-v
vmodules: /Users/wes/.vmodules
vroot: /Users/wes/dev/v
vexe: /Users/wes/dev/v/v
vexe mtime: 2022-01-13 05:17:31
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 65eb383.7882915

Git version: git version 2.30.1 (Apple Git-130)
Git vroot status: weekly.2022.02-20-g78829154
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 93fa86be

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg src/main.v`
```v
module main

fn main() {
	mut me := CatDad{}
	me.adopt<CatType,Cat>(CatType.black, BlackCat{})
}

[flag]
enum CatType {
	black
	white
}

type Cat = BlackCat | WhiteCat

struct BlackCat {}
struct WhiteCat {}

struct CatDad {}
fn (mut foo CatDad) adopt<D,T>(d D, t T) {}
```

**What did you expect to see?**

successful compilation or a regular error


**What did you see instead?**
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/main.12161416773846174658.tmp.c:10957:74: error: passing 'main__BlackCat' (aka 'struct main__BlackCat') to parameter of incompatible type 'main__Cat' (aka 'struct main__Cat')
        main__CatDad_adopt_T_main__CatType_main__Cat(&me, main__CatType__black, (main__BlackCat){EMPTY_STRUCT_INITIALIZATION});
                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/main.12161416773846174658.tmp.c:1831:113: note: passing argument to parameter 't' here
VV_LOCAL_SYMBOL void main__CatDad_adopt_T_main__CatType_main__Cat(main__CatDad* foo, main__CatType d, main__Cat t);
                                                                                                                ^
1 warning and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1471,7 +1471,37 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 	} else {
 		node.args
 	}
-	expected_types := node.expected_arg_types
+	mut expected_types := node.expected_arg_types
+	// unwrap generics fn/method arguments to concretes
+	if node.concrete_types.len > 0 && node.concrete_types.all(!it.has_flag(.generic)) {
+		if node.is_method {
+			if func := g.table.find_method(g.table.sym(node.left_type), node.name) {
+				if func.generic_names.len > 0 {
+					for i in 0 .. expected_types.len {
+						mut muttable := unsafe { &ast.Table(g.table) }
+						if utyp := muttable.resolve_generic_to_concrete(node.expected_arg_types[i],
+							func.generic_names, node.concrete_types)
+						{
+							expected_types[i] = utyp
+						}
+					}
+				}
+			}
+		} else {
+			if func := g.table.find_fn(node.name) {
+				if func.generic_names.len > 0 {
+					for i in 0 .. expected_types.len {
+						mut muttable := unsafe { &ast.Table(g.table) }
+						if utyp := muttable.resolve_generic_to_concrete(node.expected_arg_types[i],
+							func.generic_names, node.concrete_types)
+						{
+							expected_types[i] = utyp
+						}
+					}
+				}
+			}
+		}
+	}
 	// only v variadic, C variadic args will be appeneded like normal args
 	is_variadic := expected_types.len > 0 && expected_types.last().has_flag(.variadic)
 		&& node.language == .v

--- a/vlib/v/tests/generics_method_with_sumtype_args_test.v
+++ b/vlib/v/tests/generics_method_with_sumtype_args_test.v
@@ -1,0 +1,26 @@
+module main
+
+fn test_generics_method_with_sumtype_args() {
+	mut me := CatDad{}
+	ret := me.adopt<CatType, Cat>(CatType.black, BlackCat{})
+	println(ret)
+	assert ret == 22
+}
+
+[flag]
+enum CatType {
+	black
+	white
+}
+
+type Cat = BlackCat | WhiteCat
+
+struct BlackCat {}
+
+struct WhiteCat {}
+
+struct CatDad {}
+
+fn (mut foo CatDad) adopt<D, T>(d D, t T) int {
+	return 22
+}

